app/Support/Web3NftHandler.php

```diff
@@ -39,149 +40,134 @@ public function __construct(
-                $now,
-            ];
-        });
+        /** @var string[] $columns */
+        $columns = array_keys($collectionsData->first());
 
-        $valuesPlaceholders = $nftsGroupedByCollectionAddress->map(fn () => '(?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?)')->join(',');
+        unset($columns['network_id']);
+        unset($columns['address']);
 
-        $ids = DB::transaction(function () use ($nfts, $collectionsData, $valuesPlaceholders, $dispatchJobs, $now) {
-            // upsert nfts/collections (if any)
-            if ($nfts->isEmpty()) {
-                return collect();
-            }
+        // We take all the columns (except the ones we unique on which are network_id and address)
+        // and generate `column = coalesce(excluded.column, collections.column)` for UPSERT...
+        $update = collect($columns)->mapWithKeys(fn ($column) => [
+            $column => DB::raw(sprintf('coalesce(excluded.%s, collections.%s)', $column, $column)),
+        ])->all();
 
-            $dbCollections = collect(DB::select(
-                // language=postgresql
-                "
-    insert into collections
-        (address, network_id, name, slug, symbol, description, supply, floor_price, floor_price_token_id, floor_price_retrieved_at, extra_attributes, minted_block, minted_at, last_indexed_token_number, created_at, updated_at)
-    values {$valuesPlaceholders}
-    on conflict (address, network_id) do update
-        set name = trim(coalesce(excluded.name, collections.name)),
-            symbol = coalesce(excluded.symbol, collections.symbol),
-            description = coalesce(excluded.description, collections.description),
-            supply = coalesce(excluded.supply, collections.supply),
-            floor_price_token_id = excluded.floor_price_token_id,
-            floor_price_retrieved_at = excluded.floor_price_retrieved_at,
-            extra_attributes = excluded.extra_attributes,
-            minted_block = excluded.minted_block,
-            minted_at = excluded.minted_at,
-            last_indexed_token_number = coalesce(excluded.last_indexed_token_number, collections.last_indexed_token_number)
-    returning id, address, floor_price, supply
-     ",
-                $collectionsData->toArray(),
-            ));
-
-            // Group all NFTs by their address, so the key will be the address of the collection
-            $groupedByAddress = collect($dbCollections)->keyBy(fn ($nft) => Str::lower($nft->address));
-
-            $nfts = $nfts->filter(function ($nft) use ($groupedByAddress) {
-                $collection = $groupedByAddress->get(Str::lower($nft->tokenAddress));
-
-                return $this->shouldKeepNft($collection);
-            });
+        $collections = DB::transaction(function () use ($nfts, $collectionsData, $update, $now) {
+            /** @var EloquentCollection<int, CollectionModel> $collections */
+            $collections = CollectionModel::upsertReturning(
```

Comment:
Note to self,  `upsertReturning` comes from [here](https://github.com/tpetry/laravel-postgresql-enhanced/blob/b89af5de3ee36ee236525dc207f452629211c0fd/src/Eloquent/Mixins/BuilderReturning.php#L106)